### PR TITLE
fix(openapi): embed parameter modifier for search operations

### PR DESCRIPTION
**Pull request by pathec**
Hi, the response model of search operations is currently restricted to exactly one list in the `EmbedParameterModifier` when using HAL links. But regarding the [REST guide](https://sda.dev/core-concepts/communication/restful-api-guide/#RESTfulAPIGuide-SHOULD%3AProvidecomplexqueriesusingsearchresources) normally a search operation contains a list with filters that is also a part of the response besides the result list. So the filter should also handle more than one list to get the correct `embed` query parameter documented in the OpenAPI

### CODE REVIEW

**Comment by christopher-cudennec** [comment]
Hi @pathec ! Nice to see you again!

I guess you talk about this section of the REST guide the includes an example request and response, right?

```
POST /carSearches
{
    "skip": 0,
    "limit": 2,
    "filters": [
        {
            "filterType": "seatsFilter",
            "seatsMin": 4,
            "seatsMax": 10
        }
    ],
    "orderBy": ["+sign"]
}

200 OK
{
    "skip": 0,
    "limit": 2,
    "totalResults": 25,
    "filters": [
        {
            "filterType": "seatsFilter",
            "seatsMin": 4,
            "seatsMax": 10
        }
    ],
    "results": [
        {
            "id": 711,
            "manufacturer": "bmw",
            "model": "X5",
            "seats": 5,
            "sign": "HH-MM-1234"
        },
        {
            "id": 712,
            "manufacturer": "bmw",
            "model": "X5",
            "seats": 5,
            "sign": "M-CV-777"
        }
    ]
}
```

Could you add an example for the current behaviour here in the PR?
If I understand you right, the problem occurs when you add the `embed` parameter to your search query?

**Comment by pathec** [comment]
Hi @christopher-cudennec,

exactly the section in the guide I mean.

At runtime the code works fine, but the `embed` query parameter is not documented in OpenAPI currently.

The filter works fine already if the response model of an operation contains directly the HAL links, e.g. GET cars/1. But for search operations the car resource with the HAL links in defined in a list of results. Therefore the `EmbedParameterModifier` should search for HAL links in the properties of a response model and the important thing not restricted to a response model with exactly one list property.

I want for example something like this:

```yaml
  /cars:
    post:
      parameters:
      - description: "Select linked resources that should be resolved and embedded\
          \ into the response"
        in: "query"
        name: "embed"
        schema:
          type: "array"
          items:
            type: "string"
            enum:
            - "a"
            - "b"
            - "c"
     # ...
```

File: sda-commons-server-openapi/src/test/java/org/sdase/commons/server/openapi/apps/test/HouseSearchResource.java
Comment (christopher-cudennec) [inline sda-commons-server-openapi/src/test/java/org/sdase/commons/server/openapi/apps/test/HouseSearchResource.java:24]: For completeness, you should also set `this.filters = filters` in the constructor.

File: sda-commons-server-openapi/src/main/java/org/sdase/commons/optional/server/openapi/parameter/embed/EmbedParameterModifier.java
Comment (christopher-cudennec) [inline sda-commons-server-openapi/src/main/java/org/sdase/commons/optional/server/openapi/parameter/embed/EmbedParameterModifier.java:129]: Please add a `null` check to `schema.getProperties()`

File: sda-commons-server-openapi/src/test/java/org/sdase/commons/server/openapi/apps/test/Filter.java
Comment (christopher-cudennec) [inline sda-commons-server-openapi/src/test/java/org/sdase/commons/server/openapi/apps/test/Filter.java:3]: I wonder if we should extend the example to a more common use case because we usually support multiple filter types. In that case `Filter` is abstract and defines `@JsonSubTypes`. Your class would be a specific `AnimalFilter`. Any opinion, @JoergSiebahn ?

File: sda-commons-server-openapi/src/test/java/org/sdase/commons/server/openapi/apps/test/HouseSearchResource.java
Comment (pathec) [reply sda-commons-server-openapi/src/test/java/org/sdase/commons/server/openapi/apps/test/HouseSearchResource.java]: Sure, I forgot this. Done

File: sda-commons-server-openapi/src/main/java/org/sdase/commons/optional/server/openapi/parameter/embed/EmbedParameterModifier.java
Comment (pathec) [reply sda-commons-server-openapi/src/main/java/org/sdase/commons/optional/server/openapi/parameter/embed/EmbedParameterModifier.java]: Good point. I added a check

File: sda-commons-server-openapi/src/test/java/org/sdase/commons/server/openapi/apps/test/Filter.java
Comment (JoergSiebahn) [reply sda-commons-server-openapi/src/test/java/org/sdase/commons/server/openapi/apps/test/Filter.java]: As I understand the implementation and the changes (it's actually very little understanding), that detail should not affect the case that is solved in the changed modifier. I think that introducing inheritance here is just more overhead.

**Comment by christopher-cudennec** [comment]
We discussed internally that it would be nice to see the actual impact on an `openapi.yaml`. That's why I created #2734. Let's wait for this PR to be merged and rebase your changes after that. 😄

**Comment by christopher-cudennec** [comment]
@pathec : Could you please rebase your PR? 😄

**Comment by pathec** [comment]
@christopher-cudennec any blocker for this PR. I already rebased the branch twice, but before the build is ready the branch is already out-of-date :) How do you handle this?

**Comment by christopher-cudennec** [comment]
> @christopher-cudennec any blocker for this PR. I already rebased the branch twice, but before the build is ready the branch is already out-of-date :) How do you handle this?

I'll let @elbiocaetano know to pause his maintenance task and stop merging PRs until your PR was merged. 😄

**Comment by christopher-cudennec** [comment]
I wonder why the new test in [OpenApiBundleIT](https://github.com/SDA-SE/sda-dropwizard-commons/blob/1c57afb9c8344d210bbc40e854d13732c2b70980/sda-commons-server-openapi/src/test/java/org/sdase/commons/server/openapi/OpenApiBundleIT.java#L62) has not changed the expected output. I would've expected that it should fail once locally because you've changed `HouseSearchResource`. What happens when you execute it locally?

**Comment by pathec** [comment]
> I wonder why the new test in [OpenApiBundleIT](https://github.com/SDA-SE/sda-dropwizard-commons/blob/1c57afb9c8344d210bbc40e854d13732c2b70980/sda-commons-server-openapi/src/test/java/org/sdase/commons/server/openapi/OpenApiBundleIT.java#L62) has not changed the expected output. I would've expected that it should fail once locally because you've changed `HouseSearchResource`. What happens when you execute it locally?

You're right, the test fails once locally. I added the change of expected spec now to the PR. Strange that the test has passed previously

**Comment by pathec** [comment]
@christopher-cudennec  is there any manual step required for the fossa check? This check is still in progress for 3 hours

**Review by christopher-cudennec — APPROVED** [review]

**Comment by christopher-cudennec** [comment]
I will remove the check for this PR. There seems to be an issue for PRs from forks. Not sure yet what the issue is.
But since you did not change / introduce any dependencies we can safely ignore the license check.

**Comment by christopher-cudennec** [comment]
Thanks for your contribution! 😃

**Pull request merged**
